libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt

```diff
@@ -1124,4 +1219,26 @@ class MapboxRouteLineView(options: MapboxRouteLineOptions) {
         MapboxRouteLineUtils.removeLayersAndSources(style)
         MapboxRouteLineUtils.initializeLayers(style, options)
     }
+
+    private fun expectRoutes(expectedRoutes: Set<String>, callback: RoutesRenderedCallback) {
+        val renderedRoutes = mutableSetOf<String>()
+        if (expectedRoutes.isEmpty()) {
+            callback.onRoutesRendered(emptyList())
+        } else {
+            val map = callback.map
+            val listener = object : OnSourceDataLoadedListener {
+                override fun onSourceDataLoaded(eventData: SourceDataLoadedEventData) {
+                    if (eventData.loaded == true) {
+                        val routeId = sourceToFeatureMap[RouteLineSourceKey(eventData.id)]?.id()
+                        routeId?.let { renderedRoutes.add(it) }
+                        if (expectedRoutes == renderedRoutes) {
+                            map.removeOnSourceDataLoadedListener(this)
+                            callback.onRoutesRendered(renderedRoutes.toList())
```

Comment:
I see two potential issue here, first that this can be triggered by an incorrect operation and second that potential coalescing in GL-Native could prevent callbacks from being fired.

For example:
1. Call `renderRouteDrawData(longRoute1, callback1)`
2. Call `renderRouteDrawData(longRoute1withUpdatedCongestion, callback2)` before GL-Native managed to render `longRoute1`

In this situation it's very likely that `callback2` will be invoked due to the result of drawing the `longRoute1` because both have the same feature ID.

Another example is:
1. Call `renderRouteDrawData(longRoute1, callback1)`
2. Call `renderRouteDrawData(longRoute2, callback2)` before GL-Native managed to render `longRoute1`
3. Call `renderRouteDrawData(longRoute3, callback3)` before GL-Native managed to render `longRoute1` or `longRoute2`

Since the second operation is essentially a throw-away, it might as well be dropped by GL-Native which would never update the sources with the features from `longRoute2` and we'd never call back `callback2`. Perhaps we might want to consider adding a `potentiallyCanceled` (:upside_down_face:) function to `RoutesRenderedCallback` because if a new `renderRouteDrawData` is called before the previous `RoutesRenderedCallback` fired we have no guarantee that this previous update will ever be rendered.

I do think that these are edge cases though. We could mention them in the docs as known issues and move along but that's why I'd prefer the callback being directly in Maps SDK/GL-Native instead of us having to retrofit it based on the observer pattern.

Tagging @pozdnyakov @kiryldz for visibility and/or thoughts.

Reply:
I'll check the first issue.
About the second one, yeah, I like the idea of `cancelled` method. We can just invoke it from `renderRoutes` for a previous callback if it hasn't fired yet.
Surely having a Maps SDK/GL native callback-based API would simplify things.

Reply:
The more interesting case is when we do hide/show very fast or clear/draw. In this case GL Native might skip clearing/hiding the layer and we won't get notified that the routes were rendered, because they weren't cleared in the first place. And it looks like with the current API we don't have a way of understanding this on our side. And the callback will never be fired (and there's no way of telling that it was cancelled in this situation).
@pozdnyakov @kiryldz any thoughts on what we can do?

Reply:
Another problematic case:
We render route1, route2, route3.
Then we switch to an alternative and render route3, route1, route2.
`OnSourceDataLoadedListener` will fire only for route3, because others are already rendered.
It looks like without Maps/GL Native callback API we can't properly add one on NavSDK side.

Reply:
Adding DO NOT MERGE label until we decide how to deal with issues described above.

Reply:
the expected API usage implies assigning an individual id for each GeoJSON setter call for the source.
If several setter calls are performed in a row some of the intermediate calls can indeed be ignored, in that case the client is listening to the `SourceDataLoaded` event and abandons all the calls preceding the id that is passed as an event argument.

Pseudocode:
```
source.setGeoJSON("..", 1);
source.setGeoJSON("..", 2);
source.setGeoJSON("..", 3);
source.setGeoJSON("..", 4);
```
then while listening the  
```
override fun onSourceDataLoaded(eventData: SourceDataLoadedEventData) {
          if (eventData.loaded == true) {
              // eventData.id == 3 => calls 1, 2 were skipped and call 3 was complete and the corresponding data rendered
          }
      }
  }
```

Reply:
@dzinad @LukasPaczos there's a question whether the data-id is passed at all as part of this PR, from what I can see it's not (yet?). 
It should be added as a parameter to `geoJsonSource.featureCollection(geojsonData, dataId)` and will then be returned with `SourceDataLoadedEventData.dataId`. An example of usage : https://github.com/mapbox/mapbox-maps-android/blob/main/app/src/main/java/com/mapbox/maps/testapp/examples/LargeGeojsonPerformanceActivity.kt#L72, corresponding event : https://github.com/mapbox/mapbox-maps-android/blob/main/app/src/main/java/com/mapbox/maps/testapp/examples/LargeGeojsonPerformanceActivity.kt#L47.